src/controller/CommissionerDiscoveryController.cpp

```diff
@@ -135,55 +210,178 @@ void CommissionerDiscoveryController::Ok()
+                    if (target.checkState != TargetAppCheckState::kAppNotFound)
                     {
-                        // found one
-                        hasTargetApp = true;
+                        foundTargetApp = true;
                     }
                 }
-            }
-            // handle NoAppsFound CDC case
-            if (!hasTargetApp && client->GetNoPasscode())
-            {
-                ChipLogError(AppServer, "UX Ok: target apps specified but none found, sending CDC");
-                CommissionerDeclaration cd;
-                cd.SetNoAppsFound(true);
-                mUdcServer->SendCDCMessage(
-                    cd, chip::Transport::PeerAddress::UDP(client->GetPeerAddress().GetIPAddress(), client->GetCdPort()));
-                return;
+                else
+                {
+                    foundPendingTargets = true;
+                }
             }
         }
-        else
-        {
-            passcode = mPasscodeService->FetchCommissionPasscodeFromContentApp(client->GetVendorId(), client->GetProductId(),
-                                                                               rotatingIdSpan);
-        }
+    }
+    if (passcode != 0)
+    {
+        ChipLogError(AppServer, "UX Ok - HandleContentAppCheck: found a passcode");
+        // we found a passcode and complete commissioning has not been called
+        CommissionWithPasscode(passcode);
+        return;
+    }
+    if (foundPendingTargets)
+    {
+        ChipLogError(AppServer, "UX Ok - HandleContentAppCheck: have not heard from all apps");
+        // have not heard from all targets so don't do anything
+        return;
+    }
+    if (!foundTargetApp && client->GetNoPasscode())
+    {
+        // finished iterating through all apps and found none, send CDC
+        ChipLogError(AppServer, "UX Ok - HandleContentAppCheck: target apps specified but none found, sending CDC");
+        CommissionerDeclaration cd;
+        cd.SetNoAppsFound(true);
+        mUdcServer->SendCDCMessage(cd, Transport::PeerAddress::UDP(client->GetPeerAddress().GetIPAddress(), client->GetCdPort()));
+    }
+    ChipLogError(AppServer, "UX Ok - HandleContentAppCheck: advancing");
+    // otherwise, advance to the next step for trying to obtain a passcode.
+    HandleContentAppPasscodeResponse(0);
+    return;
+}
 
+/// Callback for getting execution into the main chip thread
+void CallbackHandleContentAppPasscodeResponse(System::Layer * aSystemLayer, void * aAppState)
+{
+    uint32_t * tmpPasscode = static_cast<uint32_t *>(aAppState);
+    GetCommissionerDiscoveryController()->InternalHandleContentAppPasscodeResponse(*tmpPasscode);
+    delete tmpPasscode;
+}
+
+void CommissionerDiscoveryController::HandleContentAppPasscodeResponse(uint32_t passcode)
+{
+    uint32_t * tmpPasscode = new uint32_t(passcode);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackHandleContentAppPasscodeResponse,
```

Comment:
All the same comments as above.

Reply:
added return check and removed this temporary passcode object